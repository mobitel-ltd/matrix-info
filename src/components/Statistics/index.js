import React, { useState } from 'react';
import { observer, inject } from 'mobx-react';
import moment from 'moment';
import { Table } from 'antd';

const columns = [
  {
    title: 'Проект',
    dataIndex: 'project',
    key: 'project',
    sorter: (a, b) => a.name.length - b.name.length,
    sortDirections: ['descend', 'ascend'],
  },
  {
    title: 'Наименование',
    dataIndex: 'name',
    key: 'name',
    sorter: (a, b) => {
      return a.name.localeCompare(b.name);
    },
    sortDirections: ['descend', 'ascend'],
  },
  {
    title: 'Дата изменения',
    dataIndex: 'lastEventDate',
    key: 'lastEventDate',
    sorter: (a, b) => new Date(a.lastEventDate) - new Date(b.lastEventDate),
    sortDirections: ['descend', 'ascend'],
  },
  {
    title: 'Количество пользователей',
    dataIndex: 'countMembers',
    key: 'countMembers',
    sorter: (a, b) => a.countMembers - b.countMembers,
    sortDirections: ['descend', 'ascend'],
  },
  {
    title: 'Количество сообщений',
    dataIndex: 'countMessages',
    key: 'countMessages',
    sorter: (a, b) => a.countMessages - b.countMessages,
    sortDirections: ['descend', 'ascend'],
  },
];

const title = () => 'Статистика по комнатам';

const expandedRowRender = record => (
  <ul>
    {record.members.map((member, ind) => (
      <li key={ind}>{member}</li>
    ))}
  </ul>
);

const rowSelection = setSelectedRooms => ({
  onSelect: (_, __, selectedRows) => setSelectedRooms(selectedRows),
  onSelectAll: (_, selectedRows) => setSelectedRooms(selectedRows),
});

const Statistics = ({ user: { rooms } }) => {
  // eslint-disable-next-line no-unused-vars
  const [selectedRooms, setSelectedRooms] = useState([]);
  const data = rooms.map(({ roomId, name, lastEventDate, project, messages, members }) => {
    const lastDate = moment(lastEventDate);
    return {
      key: roomId,
      name,
      members,
      countMembers: members.length,
      countMessages: messages.length,
      lastEventDate: lastDate.format('YYYY-MM-DD hh:mm'),
      project,
    };
  });
  return (
    <Table
      rowSelection={rowSelection(setSelectedRooms)}
      title={title}
      pagination={{ pageSize: 50 }}
      expandedRowRender={expandedRowRender}
      bordered
      columns={columns}
      dataSource={data}
    />
  );
};

export default inject('user')(observer(Statistics));
