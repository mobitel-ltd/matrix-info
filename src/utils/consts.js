export const baseColors = [
  '#FF6384',
  '#36A2EB',
  '#FFCE56',
  '#800000',
  '#008000',
  '#00FFFF',
  '#800080',
  '#000080',
  '#808000',
  '#FF00FF',
];
