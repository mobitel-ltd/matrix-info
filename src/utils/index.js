import { groupBy, head, last, get } from 'lodash/fp';
import url from 'url';
import { baseColors } from './consts';

const protocol = 'https';
const matrixHost = 'matrix';
export const bots = process.env.REACT_APP_BOTS.split(' ');

// export const isEnglish = val => /[\w]/.test(val);

export const getMatrixHostName = domain => [matrixHost, domain].join('.');

export const getBaseUrl = domain => url.format({ protocol, hostname: getMatrixHostName(domain) });

export const getUserId = (userName, domain) => `@${userName}:${getMatrixHostName(domain)}`;

export const getProjectName = name => {
  const [h, tail] = name.split('-');

  return tail ? h : 'custom room';
};

export const getLastEventDate = messages => last(messages) |> get('date');

const takeName = matrixId => matrixId.slice(1).split(':') |> head;

/**
 * @param {matrixSdk.Room} room matrix room
 * @returns {{name: string, roomId: string, members: string[], project: string, messages: {sender: string, date: string}[]}}
 */
export const parseRoom = room => {
  const { name } = room;
  const messages = room.timeline
    .filter(e => !e.target)
    .filter(e => !bots.some(bot => e.getSender().includes(bot)))
    .map(e => ({
      date: e.getDate(),
      sender: takeName(e.getSender()),
    }));
  const members = room.getJoinedMembers().map(({ userId }) => takeName(userId));

  return {
    name,
    messages,
    lastEventDate: getLastEventDate(messages),
    members,
    roomId: room.roomId,
    project: getProjectName(name),
  };
};

export const getRoomsUsersCount = rooms => groupBy(({ members }) => members.length, rooms) |> Object.entries;
export const getGroupedByProjects = rooms => groupBy('project', rooms) |> Object.entries;

const groupedFunc = {
  user: getRoomsUsersCount,
  project: getGroupedByProjects,
};

export const getGroupedData = type => rooms => groupedFunc[type](rooms);

export const getColors = (len, cur = 0) => {
  const currentColor = cur > baseColors.length - 1 ? 0 : cur;
  return len ? [baseColors[currentColor], ...getColors(len - 1, currentColor + 1)] : [];
};
