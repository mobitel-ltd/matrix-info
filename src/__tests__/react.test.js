import React from 'react';
import { name } from 'faker';
import { getUserId, parseRoom } from 'utils';
import renderer from 'react-test-renderer';
import Statistics from '../components/Statistics';
import { fakeRooms } from './fixtures';

let fakeMembers;
let fakeMatrixUsers;

beforeEach(() => {
  fakeMembers = [name, name, name].map(item => item.firstName().toLowerCase());
  fakeMatrixUsers = fakeMembers.map(getUserId);
});

it('render correctly text component', () => {
  const rooms = fakeRooms(fakeMatrixUsers).map(parseRoom);
  const StatisticsComponent = renderer.create(<Statistics user={{ rooms }} />).toJSON();
  expect(StatisticsComponent).toMatchSnapshot();
});
