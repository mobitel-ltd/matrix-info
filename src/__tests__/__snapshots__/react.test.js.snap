// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly text component 1`] = `
<div
  className="ant-table-wrapper"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table ant-table-default ant-table-bordered ant-table-scroll-position-left"
        style={Object {}}
      >
        <div
          className="ant-table-title"
        >
          Статистика по комнатам
        </div>
        <div
          className="ant-table-content"
        >
          <div
            className="ant-table-body"
            onScroll={[Function]}
            onWheel={[Function]}
            style={Object {}}
          >
            <table
              className=""
              style={Object {}}
            >
              <colgroup>
                <col
                  className="ant-table-expand-icon-col"
                />
                <col
                  className="ant-table-selection-col"
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
                <col
                  style={
                    Object {
                      "minWidth": undefined,
                      "width": undefined,
                    }
                  }
                />
              </colgroup>
              <thead
                className="ant-table-thead"
              >
                <tr
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <th
                    className="ant-table-expand-icon-th"
                    rowSpan={1}
                    title=""
                  />
                  <th
                    className="ant-table-selection-column"
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div>
                        <span
                          className="ant-table-column-title"
                        >
                          <div
                            className="ant-table-selection"
                          >
                            <label
                              className="ant-checkbox-wrapper"
                            >
                              <span
                                className="ant-checkbox"
                                style={Object {}}
                              >
                                <input
                                  checked={false}
                                  className="ant-checkbox-input"
                                  disabled={false}
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onFocus={[Function]}
                                  type="checkbox"
                                />
                                <span
                                  className="ant-checkbox-inner"
                                />
                              </span>
                            </label>
                          </div>
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        />
                      </div>
                    </span>
                  </th>
                  <th
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div
                        className="ant-table-column-sorters"
                      >
                        <span
                          className="ant-table-column-title"
                        >
                          Проект
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        >
                          <div
                            className="ant-table-column-sorter-inner ant-table-column-sorter-inner-full"
                            title="Sort"
                          >
                            <i
                              aria-label="icon: caret-up"
                              className="anticon anticon-caret-up ant-table-column-sorter-up off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </i>
                            <i
                              aria-label="icon: caret-down"
                              className="anticon anticon-caret-down ant-table-column-sorter-down off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </i>
                          </div>
                        </span>
                      </div>
                    </span>
                  </th>
                  <th
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div
                        className="ant-table-column-sorters"
                      >
                        <span
                          className="ant-table-column-title"
                        >
                          Наименование
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        >
                          <div
                            className="ant-table-column-sorter-inner ant-table-column-sorter-inner-full"
                            title="Sort"
                          >
                            <i
                              aria-label="icon: caret-up"
                              className="anticon anticon-caret-up ant-table-column-sorter-up off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </i>
                            <i
                              aria-label="icon: caret-down"
                              className="anticon anticon-caret-down ant-table-column-sorter-down off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </i>
                          </div>
                        </span>
                      </div>
                    </span>
                  </th>
                  <th
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div
                        className="ant-table-column-sorters"
                      >
                        <span
                          className="ant-table-column-title"
                        >
                          Дата изменения
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        >
                          <div
                            className="ant-table-column-sorter-inner ant-table-column-sorter-inner-full"
                            title="Sort"
                          >
                            <i
                              aria-label="icon: caret-up"
                              className="anticon anticon-caret-up ant-table-column-sorter-up off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </i>
                            <i
                              aria-label="icon: caret-down"
                              className="anticon anticon-caret-down ant-table-column-sorter-down off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </i>
                          </div>
                        </span>
                      </div>
                    </span>
                  </th>
                  <th
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div
                        className="ant-table-column-sorters"
                      >
                        <span
                          className="ant-table-column-title"
                        >
                          Количество пользователей
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        >
                          <div
                            className="ant-table-column-sorter-inner ant-table-column-sorter-inner-full"
                            title="Sort"
                          >
                            <i
                              aria-label="icon: caret-up"
                              className="anticon anticon-caret-up ant-table-column-sorter-up off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </i>
                            <i
                              aria-label="icon: caret-down"
                              className="anticon anticon-caret-down ant-table-column-sorter-down off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </i>
                          </div>
                        </span>
                      </div>
                    </span>
                  </th>
                  <th
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    <span
                      className="ant-table-header-column"
                    >
                      <div
                        className="ant-table-column-sorters"
                      >
                        <span
                          className="ant-table-column-title"
                        >
                          Количество сообщений
                        </span>
                        <span
                          className="ant-table-column-sorter"
                        >
                          <div
                            className="ant-table-column-sorter-inner ant-table-column-sorter-inner-full"
                            title="Sort"
                          >
                            <i
                              aria-label="icon: caret-up"
                              className="anticon anticon-caret-up ant-table-column-sorter-up off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </i>
                            <i
                              aria-label="icon: caret-down"
                              className="anticon anticon-caret-down ant-table-column-sorter-down off"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </i>
                          </div>
                        </span>
                      </div>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="2"
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDoubleClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": null,
                    }
                  }
                >
                  <td
                    className="ant-table-row-expand-icon-cell"
                  >
                    <div
                      aria-label="Expand row"
                      className="ant-table-row-expand-icon ant-table-row-collapsed"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      role="button"
                      style={Object {}}
                      tabIndex={0}
                    />
                  </td>
                  <td
                    className="ant-table-selection-column"
                    onClick={[Function]}
                  >
                    <span
                      onClick={[Function]}
                    >
                      <label
                        className="ant-checkbox-wrapper"
                      >
                        <span
                          className="ant-checkbox"
                          style={Object {}}
                        >
                          <input
                            checked={false}
                            className="ant-checkbox-input"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            type="checkbox"
                          />
                          <span
                            className="ant-checkbox-inner"
                          />
                        </span>
                      </label>
                    </span>
                  </td>
                  <td
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    MY
                  </td>
                  <td
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    MY-123
                  </td>
                  <td
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    2019-02-12 12:00
                  </td>
                  <td
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    3
                  </td>
                  <td
                    className="ant-table-column-has-actions ant-table-column-has-sorters"
                    onClick={[Function]}
                  >
                    3
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <ul
        className="ant-pagination ant-table-pagination"
        style={Object {}}
        unselectable="unselectable"
      >
        <li
          aria-disabled={true}
          className="ant-pagination-disabled ant-pagination-prev"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Previous Page"
        >
          <a
            className="ant-pagination-item-link"
          >
            <i
              aria-label="icon: left"
              className="anticon anticon-left"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </i>
          </a>
        </li>
        <li
          className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0"
          title={1}
        >
          <a>
            1
          </a>
        </li>
        <li
          aria-disabled={true}
          className="ant-pagination-disabled ant-pagination-next"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Next Page"
        >
          <a
            className="ant-pagination-item-link"
          >
            <i
              aria-label="icon: right"
              className="anticon anticon-right"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                />
              </svg>
            </i>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
